File: exercise/53/divide.cpp
class Solution {
public:
    void maxSubArrayPart(vector<int>& nums, int from, int to, int &mx, int &left, int &right, int &sum) {
        if (from == to) {
            mx = left = right = sum = nums[from];
        } else {
        int left_mx, left_left, left_right, left_sum;
        int right_mx, right_left, right_right, right_sum;
        int middle = from + (to - from) / 2;
        maxSubArrayPart(nums, from, middle, left_mx, left_left, left_right, left_sum);
        maxSubArrayPart(nums, middle + 1, to, right_mx, right_left, right_right, right_sum);
        mx = max({left_mx, right_mx, left_right + right_left});
        left = max(left_left, left_sum + right_left);
        right = max(right_right, left_right + right_sum);
        sum = left_sum + right_sum;
        }
    }

    int maxSubArray(vector<int>& nums) {
        if (nums.empty()) return 0;
        int left, right, sum, mx;
        maxSubArrayPart(nums, 0, nums.size() - 1, mx, left, right, sum);
        return mx;
    }
};

File: exercise/DP/53/divide.py
class Solution:
    def maxSubArrayPart(self, nums, left, right):
        if left == right:
            return nums[left], nums[left], nums[left], nums[left]
        mid = (left + right) // 2
        left_left, left_max, left_right, left_sum = self.maxSubArrayPart(nums, left, mid)
        right_left, right_max, right_right, right_sum = self.maxSubArrayPart(nums, mid + 1, right)
        all_left = max(left_left, left_sum + right_left)
        all_right = max(right_right, right_sum + left_right)
        all_max = max(left_max, right_max, left_right + right_left)
        all_sum = left_sum + right_sum
        return all_left, all_max, all_right, all_sum

    def maxSubArray(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        _, result, _, _ = self.maxSubArrayPart(nums, 0, len(nums) - 1)
        return result
